# Описание проекта

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

# Цель проекта
    
    Выбор региона для  добычи нефти с максимальной прибылью с учетом того, что:
    - известны показатели для оценки скважины, позволяющие предсказать возможный объем месторождения
    - всего регионов три
    - в каждом 10 000 месторождений
    - бюджет на разработку региона 10 млрд, а разработка одной скважины стоит 50 млн
    - баррель нефти приносит доход 4500
    - риск убытков должен быть не выше 2.5%.
    

# 1. Загрузка и подготовка данных

In [1]:
import pandas as pd

In [2]:
import matplotlib.pyplot as plt

In [3]:
from sklearn.linear_model import LinearRegression

In [4]:
from sklearn.model_selection import train_test_split

In [5]:
from sklearn.metrics import mean_squared_error

In [6]:
import numpy as np

In [7]:
from scipy import stats as st

 # Описание данных
Данные геологоразведки трёх регионов находятся в файлах: geo_data_0.csv, geo_data_1.csv и geo_data_2.csv.
- id — уникальный идентификатор месторождения;
- f0, f1, f2 — три признака точек (неважно, что они означают, но сами признаки значимы);
- product — объём запасов в месторождении (тыс. баррелей).

In [8]:
geo_data_0 = pd.read_csv('/datasets/geo_data_0.csv') #данные для первого региона

In [9]:
geo_data_0.head(40)

id        f0        f1        f2     product
0   txEyH  0.705745 -0.497823  1.221170  105.280062
1   2acmU  1.334711 -0.340164  4.365080   73.037750
2   409Wp  1.022732  0.151990  1.419926   85.265647
3   iJLyR -0.032172  0.139033  2.978566  168.620776
4   Xdl7t  1.988431  0.155413  4.751769  154.036647
5   wX4Hy  0.969570  0.489775 -0.735383   64.741541
6   tL6pL  0.645075  0.530656  1.780266   49.055285
7   BYPU6 -0.400648  0.808337 -5.624670   72.943292
8   j9Oui  0.643105 -0.551583  2.372141  113.356160
9   OLuZU  2.173381  0.563698  9.441852  127.910945
10  b8WQ6  0.371066 -0.036585  0.009208   70.326617
11  1YYm1  0.015920  1.062729 -0.722248   45.110381
12  zIYPq -0.276476  0.924865  0.095584   89.158678
13  iqTqq  0.212696 -0.111147  5.770095  164.298520
14  Ct5yY -0.018578  0.187516  2.944683  158.633720
15  Q77WQ  1.324566 -0.419149 -4.097429   60.677876
16  6vJcD -0.000247  0.360146  0.794115  163.143987
17  AgRR6  0.380913  1.114558  0.992377  125.942543
18  tgSSC  0.160021  0.788151  7.629164  113.793566
19  s2g8E  0.593078 -0.414964 -0.715774   28.207978
20  94tLD -1.115733  0.153296  4.068513  117.306360
21  x4axC  0.995960 -0.555728  6.539601   78.268877
22  MlCYx  0.876967 -0.524380  7.769194   94.959824
23  vYLce  1.827189 -0.131038 -0.214082   87.280632
24  RyD7G  0.279035 -0.161439  9.234609  123.126030
25  HgCfr  0.004998  1.107879  2.359406  153.998158
26  DoIei  2.166657  0.175599 -0.096555   84.540995
27  6Gdw6  0.127431  0.142624  2.789634  142.437956
28  gjrQG -0.464524  1.058439  0.356565   28.757162
29  bkKez -0.207409  1.045025 -0.430539  138.721978
30  jnQyF -1.126401  0.007046 -0.657548   87.078211
31  t86xZ  0.921726  0.405240  4.523189   89.021460
32  lSO9H -1.157504  0.362454  3.083786  131.468778
33  kkBK1  0.050464  1.130622  1.815151   67.048172
34  Gbj0z  1.703926  0.036314  4.884569   34.913885
35  ZKRFP -0.376436  0.944301  3.314538  127.494050
36  ZyXT7 -0.092077 -0.013743  6.671086   90.483554
37  TdSKv  1.133533 -0.048465  3.210587  126.838239
38  R9ocw  0.569549 -0.447580  3.114160   76.771134
39  Fo4B8  1.689161 -0.250739  3.064710  114.335445

In [10]:
geo_data_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [11]:
geo_data_1 = pd.read_csv('/datasets/geo_data_1.csv') #данные для второго региона

In [12]:
geo_data_1.head(40)

id         f0         f1        f2     product
0   kBEdx -15.001348  -8.276000 -0.005876    3.179103
1   62mP7  14.272088  -3.475083  0.999183   26.953261
2   vyE1P   6.263187  -5.948386  5.001160  134.766305
3   KcrkZ -13.081196 -11.506057  4.999415  137.945408
4   AHL4O  12.702195  -8.147433  5.004363  134.766305
5   HHckp  -3.327590  -2.205276  3.003647   84.038886
6   h5Ujo -11.142655 -10.133399  4.002382  110.992147
7   muH9x   4.234715  -0.001354  2.004588   53.906522
8   YiRkx  13.355129  -0.332068  4.998647  134.766305
9   jG6Gi   1.069227 -11.025667  4.997844  137.945408
10  7uHUe  11.777049  -5.334084  2.003033   53.906522
11  OXyvW  16.320755  -0.562946 -0.001783    0.000000
12  cB8re   7.736313  -6.093374  3.982531  107.813044
13  igmai   6.695604  -0.749449 -0.007630    0.000000
14  3OdkE -10.985487  -5.605994  2.991130   84.038886
15  3tVUO  -0.347599  -6.275884 -0.003448    3.179103
16  siveY  12.300570   2.944454  2.005541   53.906522
17  gIqzq   8.900460  -5.632857  4.994324  134.766305
18  zO5EU -13.412826  -4.729495  2.998590   84.038886
19  dkcEb  -4.373526  -8.590017  2.995379   84.038886
20  X1FZJ  -5.729511  -1.861370  3.993732  110.992147
21  eVFPS   2.195392  -3.829295  4.993895  134.766305
22  TovFu   9.391662   2.386681  5.002290  134.766305
23  KQxAV  10.852199  -3.412277  4.996318  134.766305
24  II0xj   2.581392  -2.691426  2.004898   53.906522
25  AOrbK   2.338888  -8.508383  2.008130   57.085625
26  DbtST  -0.524255  -4.685527  4.002212  110.992147
27  mqtSN   8.921872  -2.863219  0.993312   26.953261
28  dqN1d  10.373859   1.145210  4.000138  107.813044
29  JPiGi -10.926274  -4.017907  0.005142    3.179103
30  YVVPX  11.726927  -4.285120  1.003547   26.953261
31  4NJv3  -7.736177  -4.857165  2.995679   84.038886
32  2DKVI  10.209863 -13.659814  5.000668  137.945408
33  M57PU  -2.286528 -12.063964  2.007202   57.085625
34  5O3Zt  -4.792992  -7.569682  1.004281   30.132364
35  BZWmy   0.695137 -10.214670  2.004539   53.906522
36  aArVp   5.897063  -5.249982  2.994434   80.859783
37  EHH3V  -4.302352  -1.846049  4.000282  110.992147
38  fNWXc  -5.769301  -5.648602  4.000441  110.992147
39  957X9   6.037017   2.572420  2.001547   53.906522

In [13]:
geo_data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [14]:
geo_data_2 = pd.read_csv('/datasets/geo_data_2.csv') #данные для третьего региона

In [15]:
geo_data_2.head(40)

id        f0        f1        f2     product
0   fwXo0 -1.146987  0.963328 -0.828965   27.758673
1   WJtFt  0.262778  0.269839 -2.530187   56.069697
2   ovLUW  0.194587  0.289035 -5.586433   62.871910
3   q6cA6  2.236060 -0.553760  0.930038  114.572842
4   WPMUX -0.515993  1.716266  5.899011  149.600746
5   LzZXx -0.758092  0.710691  2.585887   90.222465
6   WBHRv -0.574891  0.317727  1.773745   45.641478
7   XO8fn -1.906649 -2.458350 -0.177097   72.480640
8   ybmQ5  1.776292 -0.279356  3.004156  106.616832
9   OilcN -1.214452 -0.439314  5.922514   52.954532
10  myq2g  0.628659 -2.120049  5.362437  166.799132
11  V9R1y -0.544565  0.039088  8.635293  150.859435
12  bKDo0 -0.960453 -0.127113  7.764734   96.910855
13  eXvhs  1.432891 -1.242074  3.557168    8.704162
14  wI4j3 -0.935464  2.420523  1.150536  116.829811
15  XOWTx -2.628353  0.937303  0.057847   68.655498
16  2oVFv -0.428050 -2.246556  6.970420   84.500464
17  0wbDA  0.735307  0.913124  7.899539   96.284701
18  NhHaR  0.706261 -0.993553  3.155582   48.052268
19  DMNMZ -1.867846  2.744144  5.918278  153.815137
20  3mCRf -1.250704  0.540611  3.438567   65.774963
21  fRllb -0.307564 -0.942793  1.571509  113.080584
22  6GvzI -1.221304  1.262642  5.198435  105.207088
23  JwbXn  0.426604  1.142824  6.360318  108.375984
24  urwUC  2.378418  0.665386  1.282287   56.478241
25  KMQl3  0.384287 -2.416239 -1.111076   95.466896
26  Coy8M  0.927511  0.769120  3.512492   94.251843
27  EXElh -0.387211  0.160795  4.097400   77.404287
28  uaZJK  0.513782  1.102372 -0.305253   25.626312
29  6UkRY  1.043879 -0.570642  0.751992  103.263475
30  PMtzo  1.749419  2.974820  1.663992  145.730271
31  y3a1h -1.764716  1.175664  2.898374  118.130136
32  wBJrC -0.263935 -0.442263  3.326272   40.285336
33  KLqMx -0.452600 -2.028847 -5.497120   29.861027
34  HE3Rv  1.903590  1.077743 -1.622128   66.334870
35  FAoc2  1.320987 -1.343799  0.935470  107.040058
36  dQ5QX -1.231645  0.038604  3.394944   49.949507
37  FmeFh  1.549409 -0.263697  2.151824  112.645805
38  nja19 -1.364640  3.167210  6.533984   62.207685
39  I0A8A -1.423237 -3.503145  0.912362  109.513921

In [16]:
geo_data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


Нормируемыми показателями качества товарной нефти являются:

- содержание в ней воды;
- количество примесей механического характера;
- показатель давления насыщенных нефтяных паров;
- содержание солей хлористого вида;
- содержание хлор – органических химических соединений.

Эти показатели не могут быть отрицательными, следовательно, нужно взять их модули (вероятно, внесли ошибочно отрицательные значения).
Также переведем все названия в нижний регистр и проверим на дубликаты.

Напишем функцию, т.к. у нас 3 датасета.

In [17]:
def predobrab(data):
    data[['f0', 'f1', 'f2', 'product']] = data[['f0', 'f1', 'f2', 'product']].abs()
    data['id'] = data['id'].str.lower()
    return data

In [18]:
geo_data_0 = predobrab(geo_data_0) # проверяем новый вид датасета, а также количество дубликатов (аналогично - для двух других датасетов) 
print(geo_data_0.head(),'\n', 'Количество дубликатов:', geo_data_0.duplicated().sum())

      id        f0        f1        f2     product
0  txeyh  0.705745  0.497823  1.221170  105.280062
1  2acmu  1.334711  0.340164  4.365080   73.037750
2  409wp  1.022732  0.151990  1.419926   85.265647
3  ijlyr  0.032172  0.139033  2.978566  168.620776
4  xdl7t  1.988431  0.155413  4.751769  154.036647 
 Количество дубликатов: 0


In [19]:
geo_data_1 = predobrab(geo_data_1) 
print(geo_data_1.head(),'\n', 'Количество дубликатов:', geo_data_1.duplicated().sum())

      id         f0         f1        f2     product
0  kbedx  15.001348   8.276000  0.005876    3.179103
1  62mp7  14.272088   3.475083  0.999183   26.953261
2  vye1p   6.263187   5.948386  5.001160  134.766305
3  kcrkz  13.081196  11.506057  4.999415  137.945408
4  ahl4o  12.702195   8.147433  5.004363  134.766305 
 Количество дубликатов: 0


In [20]:
geo_data_2 = predobrab(geo_data_2) 
print(geo_data_2.head(),'\n', 'Количество дубликатов:', geo_data_2.duplicated().sum())

      id        f0        f1        f2     product
0  fwxo0  1.146987  0.963328  0.828965   27.758673
1  wjtft  0.262778  0.269839  2.530187   56.069697
2  ovluw  0.194587  0.289035  5.586433   62.871910
3  q6ca6  2.236060  0.553760  0.930038  114.572842
4  wpmux  0.515993  1.716266  5.899011  149.600746 
 Количество дубликатов: 0


# 2. Обучение и проверка модели

Предполагаем, что product (объём запасов в месторождения) зависит от трех показателей(f0, f1, f2).
В качестве модели используется линейная регрессия (не склонна к переобучению из-за малого количества параметров). Работать будем со всеми 100 000 месторождениями (чем больше значений, тем лучше качество модели).

Разделяем исходные данные на обучающую и валидационную выборки в соотношении 75:25.
Напишем функцию, т.к. у нас 3 датасета.

In [27]:
def data_split(data): # функция разделения данных на обучающую и валидационную выборки 
    target = data['product']
    features = data.drop(columns=['product', 'id'] , axis=1)
    #Чтобы не мешать модели, удалим и столбец, который не несет нужной нам информации.
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25, random_state=12345)
    return features_train, features_valid, target_train, target_valid

In [28]:
features_train0, features_valid0, target_train0, target_valid0 = data_split(geo_data_0)

In [29]:
features_train1, features_valid1, target_train1, target_valid1 = data_split(geo_data_1)

In [30]:
features_train2, features_valid2, target_train2, target_valid2 = data_split(geo_data_2)

In [31]:
def model(features_train, features_valid, target_train, target_valid):
# функция для работы модели линейной регрессии, ниже выведем на экран для трех регионов показатели
    model = LinearRegression()
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    mse = mean_squared_error(target_valid, predictions_valid)
    rmse = mse ** 0.5
    product_mean = predictions_valid.mean()
    return product_mean, rmse

In [32]:
print(' Cредний запас сырья и RMSE модели для датасета geo_data_0', model(features_train0, features_valid0, target_train0, target_valid0), 'соответственно.','\n', 'Cредний запас сырья и RMSE модели для датасета geo_data_1', model(features_train1, features_valid1, target_train1, target_valid1), 'соответственно.', '\n', 'Cредний запас сырья и RMSE модели для датасета geo_data_2', model(features_train2, features_valid2, target_train2, target_valid2), 'соответственно.')

 Cредний запас сырья и RMSE модели для датасета geo_data_0 (92.51881935132751, 40.3427566762566) соответственно. 
 Cредний запас сырья и RMSE модели для датасета geo_data_1 (68.73031469768347, 1.5360594290776344) соответственно. 
 Cредний запас сырья и RMSE модели для датасета geo_data_2 (95.03696840612521, 40.615887208944) соответственно.


In [34]:
def predictions(features_train, features_valid, target_train, target_valid):
# функция для создания списка с предсказаниями объемов запасов
    model = LinearRegression()
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    return predictions_valid

In [35]:
predictions_valid0 = predictions(features_train0, features_valid0, target_train0, target_valid0)

In [36]:
predictions_valid2 = predictions(features_train2, features_valid2, target_train2, target_valid2)

In [37]:
geo_data_0['predictions'] = pd.DataFrame(predictions_valid0 ) #добавляем столбец с предсказаниями для двух таблиц отобранных регионов

In [38]:
geo_data_2['predictions'] = pd.DataFrame(predictions_valid2 )

In [39]:
geo_data_0.head()

id        f0        f1        f2     product  predictions
0  txeyh  0.705745  0.497823  1.221170  105.280062    91.641670
1  2acmu  1.334711  0.340164  4.365080   73.037750    75.046616
2  409wp  1.022732  0.151990  1.419926   85.265647    81.217538
3  ijlyr  0.032172  0.139033  2.978566  168.620776    73.260562
4  xdl7t  1.988431  0.155413  4.751769  154.036647    64.032154

In [40]:
geo_data_0 = geo_data_0.dropna() 
# т.к. предсказания были только для части датасетов, удаляем строки нез предсказаний. У нас останется 25000 строк - это достаточно для оценки прибыли.

In [41]:
geo_data_2 = geo_data_2.dropna()

In [42]:
geo_data_0.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25000 entries, 0 to 24999
Data columns (total 6 columns):
id             25000 non-null object
f0             25000 non-null float64
f1             25000 non-null float64
f2             25000 non-null float64
product        25000 non-null float64
predictions    25000 non-null float64
dtypes: float64(5), object(1)
memory usage: 1.3+ MB


# 3. Подготовка к расчёту прибыли

Для расчета прибыли нам понабятся следующие данные (константы Python):

In [49]:
n = 500 # количество точек для исследования
b = 10000000000 # бюджет на разработку месторождений 
w = 50000000 # стоимость бурения одной скважины
br = 4500 # прибыль одного барреля сырья 

Бюджет на разработку месторождений 10 млрд.
Чтобы получить выгоду, нам необходимо выручить больше данной суммы с одного месторождения.

Исходя из бюджета на разработку месторождений — 10 млрд рублей и стоимость бурения одной скважины — 50 млн рублей, на этот бюджет мы сможем пробурить 200 скважин (test_n , далее будем использовать эту цифру для выборки).

Посчитаем:
    - 10000000000 / 4500 = 2 222 222 - суммарное искомое кол-во баррелей, которые нужно получить (минимальный средний объём сырья в месторождениях региона, достаточный для его разработки)
    - 2 222 222 / 200 = 11111 - минимальное среднее количество барелей, которое должна принести каждая скважина.
    
В отобранных регионах мы имеем средний объём сырья для одной скважины для geo_data_0 -  92519 баррелей и для geo_data_2 -  95037 баррелей.

Этого объема должно быть достаточно.

In [50]:
test_n = 200 # Максимальное кол-во скважин, которое можно пробурить, исходя их бюджета

In [51]:
def revenue(probabilities): # функция для расчёта прибыли по набору отобранных месторождений и предсказаний модели
    return test_n * br * probabilities*1000 - b

# 4. Расчёт прибыли и рисков 

In [52]:
state = np.random.RandomState(12345)

In [53]:
def camputation(data):
# функция для расчета средняя прибыли и 95%-й доверительный интервала с использование техники Bootstrap с 1000 выборок, чтобы найти распределение прибыли 
    values = []

    for i in range(1000):
        target_subsample = data['predictions'].sample(n=200, replace=True, random_state=state)
        values.append(revenue(target_subsample.mean()*1000))

    values = pd.Series(values)

    mean = values.mean()
    confidence_interval = st.t.interval(0.95, len(values)-1, values.mean(), values.sem())
    return  mean, confidence_interval

In [54]:
print(' Cредняя прибыль и 95%-й доверительный интервал для первого региона:', '\n', camputation(geo_data_0), '\n','Cредняя прибыль и 95%-й доверительный интервал для третьего региона:', '\n', camputation(geo_data_2))

 Cредняя прибыль и 95%-й доверительный интервал для первого региона: 
 (83255857009041.25, (83183632390049.83, 83328081628032.67)) 
 Cредняя прибыль и 95%-й доверительный интервал для третьего региона: 
 (85508458203597.25, (85434277177267.1, 85582639229927.4))


✔Вывод: средняя прибыль третьего региона больше первого на 2,7% или 2285 млрд.

In [55]:
def rev_values(data): # функция для создания списка средних значений прибыли
    values = []

    for i in range(1000):
        target_subsample = data['predictions'].sample(n=200, replace=True, random_state=state)
        values.append(revenue(target_subsample.mean()*1000))

    values = pd.Series(values)

    return  values

Напишем функцию для гипотезы: средняя прибыль меньше бюджета разработки (b=10 млрд),  альтернативная гипотеза - средняя прибыль не меньше бюджета разработки

In [56]:
def risks(data):
    alpha = 0.05 # критический уровень статистической значимости для оценки 2.5% риска

    # тест односторонний: p-value будет в два раза меньше
    
    results = st.ttest_1samp(rev_values(data), b) 
    pvalue = (results.pvalue) / 2
    
# отвергаем гипотезу и только если выборочное среднее значимо меньше бюджета

    if (results.pvalue < alpha) and (rev_values(data).mean() < b):
        answer = "Риск превышает 2.5%"
    else:
        answer = "Риск не превышает 2.5% "
    return answer

In [57]:
print(' Для первого региона:', '\n', risks(geo_data_0), '\n', 'Для третьего региона:', '\n', risks(geo_data_2)) 

 Для первого региона: 
 Риск не превышает 2.5%  
 Для третьего региона: 
 Риск не превышает 2.5% 


# Общие выводы


В задече нам требовалось выбрать регион для  добычи нефти с максимальной прибылью с учетом того, что:
- известны показатели для оценки скважины, позволяющие предсказать возможный объем месторождения
- всего регионов три
- в каждом 10 000 месторождений
- бюджет на разработку региона 10 млрд, а разработка одной скважины стоит 50 млн
- баррель нефти приносит доход 4500
- риск убытков должен быть не выше 2.5%.

В ходе проделанного исследования были выбраны 1 и 3 регионы месторождений для дальнейшего изучения (со средними запасами 92.52 и 95.04 соответственно), несмотря на значительно худшие показатели RMSE, чем во втором регионе (самуя точная модель (RMSE=1.54), однако среднее значение запасов тут самое малое (68.73)).

При расчетах получили следующие значения:

     Cредняя прибыль и 95%-й доверительный интервал для первого региона: 
     83233146701798.36, (83161609787138.92, 83304683616457.8)
     Cредняя прибыль и 95%-й доверительный интервал для третьего региона: 
     85518757154029.72, (85444970873780.17, 85592543434279.27)

Риск убытков не превышает 2.5% для обоих регионов.

При сравнении полученных данных прибыль третьего региона больше первого на 2,7% или 2285 млрд, поэтому для разработки предлагается третий регион.
